File: src/main/java/ru/arc/misc/BaltopGui.java
package ru.arc.misc;

import java.util.Comparator;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.stream.Collectors;

import com.github.stefvanschie.inventoryframework.adventuresupport.TextHolder;
import com.github.stefvanschie.inventoryframework.gui.GuiItem;
import com.github.stefvanschie.inventoryframework.gui.type.ChestGui;
import com.github.stefvanschie.inventoryframework.pane.OutlinePane;
import com.github.stefvanschie.inventoryframework.pane.PaginatedPane;
import com.github.stefvanschie.inventoryframework.pane.Pane;
import com.github.stefvanschie.inventoryframework.pane.StaticPane;
import lombok.SneakyThrows;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import ru.arc.ARC;
import ru.arc.configs.Config;
import ru.arc.hooks.HookRegistry;
import ru.arc.util.CooldownManager;
import ru.arc.util.GuiUtils;
import ru.arc.util.ItemStackBuilder;
import ru.arc.util.TextUtil;

import static ru.arc.util.Logging.info;
import static ru.arc.util.TextUtil.formatAmount;
import static ru.arc.util.TextUtil.mm;

public class BaltopGui extends ChestGui {

    record BaltopGuiItem(GuiItem item, double balance, double bank, double total) {
    }

    private static List<BaltopGuiItem> cachedItems = new CopyOnWriteArrayList<>();
    private static long lastUpdate = 0;

    final Config config;
    final Player player;

    int rows;

    GuiItem nextItem, prevItem, sortItem;
    PaginatedPane pane;
    Sort sort = Sort.TOTAL;

    public BaltopGui(Config config, Player player) {
        super(6, TextHolder.deserialize(
                TextUtil.toLegacy(config.string("baltop.title"))
        ));
        this.config = config;
        this.player = player;

        rows = config.integer("baltop.rows", 6);
        setRows(rows);

        setupBackground();
        setupPlayers();
        setupNavigation();
    }

    private void setupNavigation() {
        StaticPane nav = new StaticPane(0, rows - 1, 9, 1);
        nextItem = new ItemStackBuilder(Material.BLUE_STAINED_GLASS_PANE)
                .modelData(11008)
                .display(config.string("baltop.next.name"))
                .lore(config.stringList("baltop.next.lore"))
                .toGuiItemBuilder()
                .clickEvent(click -> {
                    click.setCancelled(true);
                    if (pane.getPage() == pane.getPages() - 1) return;
                    pane.setPage(pane.getPage() + 1);
                    this.update();
                }).build();

        prevItem = new ItemStackBuilder(Material.BLUE_STAINED_GLASS_PANE)
                .modelData(11009)
                .display(config.string("baltop.previous.name"))
                .lore(config.stringList("baltop.previous.lore"))
                .toGuiItemBuilder()
                .clickEvent(click -> {
                    click.setCancelled(true);
                    if (pane.getPage() == 0) return;
                    pane.setPage(pane.getPage() - 1);
                    this.update();
                }).build();

        sortItem = new ItemStackBuilder(Material.BLUE_STAINED_GLASS_PANE)
                .modelData(11021)
                .display(config.string("baltop.sort.name")
                        .replace("<sort>", switch (sort) {
                            case BALANCE -> config.string("baltop.sort.balance");
                            case BANK -> config.string("baltop.sort.bank");
                            case TOTAL -> config.string("baltop.sort.total");
                        }))
                .lore(config.stringList("baltop.sort.lore").stream()
                        .map(s -> s.replace("<sort>", switch (sort) {
                            case BALANCE -> config.string("baltop.sort.balance");
                            case BANK -> config.string("baltop.sort.bank");
                            case TOTAL -> config.string("baltop.sort.total");
                        })).toList())
                .toGuiItemBuilder()
                .clickEvent(click -> {
                    click.setCancelled(true);

                    if (CooldownManager.cooldown(player.getUniqueId(), "baltop_sort") != 0) {
                        GuiUtils.temporaryChange(sortItem.getItem(),
                                mm(config.string("baltop.sort.cooldown"), true),
                                null, 20 * 3, this::update);
                        return;
                    }

                    CooldownManager.addCooldown(player.getUniqueId(), "baltop_sort", 1000);

                    switch (sort) {
                        case BALANCE -> sort = Sort.BANK;
                        case BANK -> sort = Sort.TOTAL;
                        case TOTAL -> sort = Sort.BALANCE;
                    }
                    Bukkit.getScheduler().runTaskAsynchronously(ARC.plugin, () -> {
                        setupPlayers();
                        this.update();
                    });
                }).build();


        GuiItem back = new ItemStackBuilder(Material.BLUE_STAINED_GLASS_PANE)
                .modelData(11013)
                .display(config.string("baltop.back.name"))
                .lore(config.stringList("baltop.back.lore"))
                .toGuiItemBuilder()
                .clickEvent(click -> {
                    click.setCancelled(true);
                    player.closeInventory();
                    player.performCommand(config.string("baltop.back-command"));
                }).build();
        nav.addItem(nextItem, 5, 0);
        nav.addItem(prevItem, 3, 0);
        nav.addItem(sortItem, 4, 0);
        nav.addItem(back, 0, 0);
        this.addPane(nav);
    }

    @SneakyThrows
    private void setupPlayers() {
        var future = updateCache(this::update);
        if (!Bukkit.isPrimaryThread()) future.get();
        pane = new PaginatedPane(0, 0, 9, rows - 1);
        Comparator<BaltopGuiItem> comparator = switch (sort) {
            case BALANCE -> Comparator.comparingDouble(BaltopGuiItem::balance).reversed();
            case BANK -> Comparator.comparingDouble(BaltopGuiItem::bank).reversed();
            default -> Comparator.comparingDouble(BaltopGuiItem::total).reversed();
        };
        pane.populateWithGuiItems(cachedItems.stream().sorted(comparator).map(BaltopGuiItem::item).toList());
        this.addPane(pane);
    }


    private CompletableFuture<Void> updateCache(Runnable callback) {
        if (HookRegistry.redisEcoHook == null) return CompletableFuture.completedFuture(null);
        if (System.currentTimeMillis() - lastUpdate <= 60000) return CompletableFuture.completedFuture(null);
        info("Updating baltop cache");
        return HookRegistry.redisEcoHook.getTopAccounts(224)
                .thenAccept(accounts -> {
                    record Account(String name, UUID uuid, double balance, double bank) {
                    }
                    var list = accounts.stream().map(account -> {
                                double bank = HookRegistry.bankHook == null ? 0 :
                                        HookRegistry.bankHook.offlineBalance(account.uuid.toString());
                                return new Account(account.name, account.uuid,
                                        account.balance + bank, bank);
                            })
                            .filter(account -> account.balance() > 0.0)
                            //.sorted(Comparator.comparing(Account::balance).reversed())
                            .map(account -> generateGuiitem(account.name(), account.uuid(), account.balance(), account.bank()))
                            .toList();
                    cachedItems.clear();
                    cachedItems.addAll(list);
                }).thenAccept(v -> callback.run());
    }

    private BaltopGuiItem generateGuiitem(String name, UUID uuid, double total, double bank) {
        GuiItem guiItem = new ItemStackBuilder(Material.PLAYER_HEAD)
                .skull(uuid)
                .display(config.string("baltop.item.name")
                        .replace("<player>", name)
                        .replace("<balance>", formatAmount(total - bank))
                        .replace("<total>", formatAmount(total))
                        .replace("<bank>", formatAmount(bank)))
                .lore(config.stringList("baltop.item.lore").stream()
                        .map(s -> s.replace("<balance>", formatAmount(total - bank))
                                .replace("<bank>", formatAmount(bank))
                                .replace("<total>", formatAmount(total))
                                .replace("<player>", name))
                        .collect(Collectors.toList()))
                .toGuiItemBuilder()
                .clickEvent(click -> click.setCancelled(true))
                .build();
        return new BaltopGuiItem(guiItem, total - bank, bank, total);
    }

    private void setupBackground() {
        OutlinePane pane = new OutlinePane(0, rows - 1, 9, 1);
        pane.addItem(GuiUtils.background());
        pane.setRepeat(true);
        pane.setPriority(Pane.Priority.LOWEST);
        this.addPane(pane);

        OutlinePane pane2 = new OutlinePane(0, 0, 9, rows - 1);
        pane2.addItem(
                new ItemStackBuilder(Material.GRAY_STAINED_GLASS_PANE)
                        .display(" ")
                        .toGuiItemBuilder()
                        .clickEvent(click -> {
                            click.setCancelled(true);
                        }).build()
        );
        pane2.setRepeat(true);
        pane2.setPriority(Pane.Priority.LOWEST);
        this.addPane(pane2);
    }


    enum Sort {
        BALANCE, BANK, TOTAL
    }
}

File: src/main/kotlin/ru/arc/misc/BaltopGui.kt
package ru.arc.misc

import net.kyori.adventure.text.Component
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.inventory.meta.SkullMeta
import ru.arc.config.Config
import ru.arc.core.sync
import ru.arc.gui.ArcMenuSchema
import ru.arc.gui.ArcMenus
import ru.arc.hooks.HookRegistry
import ru.arc.paper.menu.PaperMenuItemRenderContext
import ru.arc.util.CooldownManager
import ru.arc.util.Logging.error
import ru.arc.util.Logging.info
import ru.arc.util.TextUtil
import ru.arc.util.TextUtil.formatAmount
import ru.arc.util.TextUtil.mm
import java.util.UUID
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Balance leaderboard GUI.
 */
object BaltopGuiFactory {
    private data class BaltopEntry(
        val name: String,
        val uuid: UUID,
        val balance: Double,
        val bank: Double,
    ) {
        val total: Double get() = balance + bank
    }

    enum class Sort { BALANCE, BANK, TOTAL }

    private val cachedEntries = CopyOnWriteArrayList<BaltopEntry>()
    private var lastUpdate = 0L
    private var refreshInFlight: CompletableFuture<Void>? = null

    fun open(
        config: Config,
        player: Player,
        sort: Sort = Sort.TOTAL,
    ) {
        val comparator: Comparator<BaltopEntry> =
            when (sort) {
                Sort.BALANCE -> compareByDescending { it.balance }
                Sort.BANK -> compareByDescending { it.bank }
                Sort.TOTAL -> compareByDescending { it.total }
            }
        val sortedEntries = cachedEntries.sortedWith(comparator)
        val entries = sortedEntries.map { entry ->
            val item = ArcMenus.item(
                "baltop-entry",
                PaperMenuItemRenderContext(values = mapOf(
                    "player" to Component.text(entry.name),
                    "balance" to Component.text(formatAmount(entry.balance)),
                    "bank" to Component.text(formatAmount(entry.bank)),
                    "total" to Component.text(formatAmount(entry.total)),
                )),
            )
            (item.itemMeta as? SkullMeta)?.let { meta ->
                meta.owningPlayer = Bukkit.getOfflinePlayer(entry.uuid)
                item.itemMeta = meta
            }
            ArcMenus.entry(item, enabled = false)
        }
        val sortName = when (sort) {
            Sort.BALANCE -> config.string("baltop.sort.balance")
            Sort.BANK -> config.string("baltop.sort.bank")
            Sort.TOTAL -> config.string("baltop.sort.total")
        }
        ArcMenus.open(
            player,
            ArcMenuSchema.BALTOP,
            config.component("baltop.title", "<dark_gray>Топ богачей"),
            elements = mapOf(
                "back" to ArcMenus.entry(ArcMenus.item(ArcMenuSchema.BALTOP, "back")) { it.closeInventory() },
                "previous" to ArcMenus.entryWithContext(ArcMenus.item(ArcMenuSchema.BALTOP, "previous")) { it.session.previousPage() },
                "sort" to ArcMenus.entry(ArcMenus.item(ArcMenuSchema.BALTOP, "sort", PaperMenuItemRenderContext(
                    values = mapOf("sort" to TextUtil.mm(sortName, true)),
                ))) { clicker ->
                    if (CooldownManager.cooldown(clicker.uniqueId, "baltop_sort") != 0L) {
                        clicker.sendMessage(mm(config.string("baltop.sort.cooldown"), true))
                    } else {
                        CooldownManager.addCooldown(clicker.uniqueId, "baltop_sort", 1000L)
                        val next = when (sort) {
                            Sort.BALANCE -> Sort.BANK
                            Sort.BANK -> Sort.TOTAL
                            Sort.TOTAL -> Sort.BALANCE
                        }
                        open(config, clicker, next)
                    }
                },
                "next" to ArcMenus.entryWithContext(ArcMenus.item(ArcMenuSchema.BALTOP, "next")) { it.session.nextPage() },
            ),
            regions = mapOf(ArcMenuSchema.BALTOP_ENTRIES to entries),
        )
    }

    /**
     * Update cache if needed and show GUI.
     */
    fun showAsync(
        config: Config,
        player: Player,
        sort: Sort = Sort.TOTAL,
    ) {
        updateCacheIfNeeded()
            .thenRun {
                sync { open(config, player, sort) }
            }.exceptionally { failure ->
                error("Failed to show baltop for {}", player.name, failure.cause ?: failure)
                sync {
                    player.sendMessage(
                        mm(
                            config.string(
                                "baltop.error",
                                "<red>Не удалось загрузить таблицу лидеров. Попробуйте ещё раз позже.",
                            ),
                            true,
                        ),
                    )
                }
                null
            }
    }

    /**
     * Update the cache if it's stale.
     */
    @Synchronized
    private fun updateCacheIfNeeded(): CompletableFuture<Void> {
        val redisEco = HookRegistry.redisEcoHook
        if (redisEco == null) {
            return CompletableFuture.completedFuture(null)
        }

        if (System.currentTimeMillis() - lastUpdate <= 60000) {
            return CompletableFuture.completedFuture(null)
        }
        refreshInFlight?.takeUnless { it.isDone }?.let { return it }

        info("Updating baltop cache")
        val refresh =
            redisEco
            .getTopAccounts(224)
            .thenAccept { accounts ->
                val entries =
                    accounts
                        .mapNotNull { account ->
                            val name = account.name ?: return@mapNotNull null
                            val uuid = account.uuid ?: return@mapNotNull null
                            val bank = HookRegistry.bankHook?.offlineBalance(uuid.toString()) ?: 0.0
                            BaltopEntry(
                                name = name,
                                uuid = uuid,
                                balance = account.balance,
                                bank = bank,
                            )
                        }.filter { it.total > 0.0 }

                cachedEntries.clear()
                cachedEntries.addAll(entries)
                lastUpdate = System.currentTimeMillis()
            }
        refreshInFlight = refresh
        refresh.whenComplete { _, _ ->
            synchronized(this) {
                if (refreshInFlight === refresh) {
                    refreshInFlight = null
                }
            }
        }
        return refresh
    }
}
